zerver/lib/retention.py

```diff
@@ -300,17 +294,120 @@ def archive_messages(chunk_size: int=MESSAGE_BATCH_SIZE) -> None:
+                  archive_transaction_id=archive_transaction_id,
+                  archive_table_name=model['archive_table_name'])
+
+def restore_attachments_from_archive(archive_transaction_id: int) -> None:
+    query = """
+        INSERT INTO zerver_attachment ({dst_fields})
+        SELECT {src_fields}
+        FROM zerver_archivedattachment
+        INNER JOIN zerver_archivedattachment_messages
+            ON zerver_archivedattachment_messages.archivedattachment_id = zerver_archivedattachment.id
+        INNER JOIN zerver_archivedmessage
+            ON  zerver_archivedattachment_messages.archivedmessage_id = zerver_archivedmessage.id
+        LEFT JOIN zerver_attachment ON zerver_archivedattachment.id = zerver_attachment.id
+        WHERE zerver_archivedmessage.archive_transaction_id = {archive_transaction_id}
+            AND zerver_attachment.id IS NULL
+        GROUP BY zerver_archivedattachment.id
+    """
+    move_rows(Attachment, query, src_db_table='zerver_archivedattachment',
+              archive_transaction_id=archive_transaction_id)
+
+def restore_attachment_messages_from_archive(archive_transaction_id: int) -> None:
+    query = """
+        INSERT INTO zerver_attachment_messages (id, attachment_id, message_id)
+        SELECT zerver_archivedattachment_messages.id,
+            zerver_archivedattachment_messages.archivedattachment_id,
+            zerver_archivedattachment_messages.archivedmessage_id
+        FROM zerver_archivedattachment_messages
+        INNER JOIN zerver_archivedmessage
+            ON  zerver_archivedattachment_messages.archivedmessage_id = zerver_archivedmessage.id
+        LEFT JOIN zerver_attachment_messages
+            ON zerver_archivedattachment_messages.id = zerver_attachment_messages.id
+        WHERE zerver_archivedmessage.archive_transaction_id = {archive_transaction_id}
+                AND  zerver_attachment_messages.id IS NULL
+    """
+    with connection.cursor() as cursor:
+        cursor.execute(query.format(archive_transaction_id=archive_transaction_id))
+
+@transaction.atomic
+def restore_data_from_archive(archive_transaction: ArchiveTransaction) -> None:
+    restore_messages_from_archive(archive_transaction.id)
+    restore_models_with_message_key_from_archive(archive_transaction.id)
+    restore_attachments_from_archive(archive_transaction.id)
+    restore_attachment_messages_from_archive(archive_transaction.id)
+
+    archive_transaction.restored = True
+    archive_transaction.save()
+
+def restore_data_from_archive_by_transactions(archive_transactions: List[ArchiveTransaction]) -> None:
+    # Looping over the list of ids means we're batching the restoration process by the size of the
+    # transactions:
+    for archive_transaction in archive_transactions:
+        restore_data_from_archive(archive_transaction)
+
+def restore_data_from_archive_by_realm(realm: Realm) -> None:
+    transactions = ArchiveTransaction.objects.exclude(restored=True).filter(realm=realm)
```

Comment:
Shouldn't this also be filtering on `type=ArchiveTransaction.RETENTION_POLICY_BASED` ?

Reply:
Not strictly necessary, because as of now, a transaction has non-NULL realm if and only if the type is RETENTION_POLICY_BASED. But perhaps filtering the type is clearer and safer for the future.

Reply:
Yes, especially since we'll likely decide we want to add a `realm` attribute to those other transactions (probably the realm of the user doing the deletion).